006-vortexa-tech-challenge/src/map.tsx

```diff
@@ -0,0 +1,87 @@
+import "maplibre-gl/dist/maplibre-gl.css";
+
+import { Feature, MultiPolygon } from "geojson";
+import { RampData, RampProperties } from "./types";
+import React, { Dispatch, SetStateAction, useState } from "react";
+import ReactMap, { Marker, Popup, ViewStateChangeEvent } from "react-map-gl";
+
+import maplibregl from "maplibre-gl";
+
+export const Map = ({
+  ramps,
+  visibleFeatures,
+  setVisibleFeatures,
+}: {
+  ramps: RampData;
+  visibleFeatures: Feature<MultiPolygon, RampProperties>[] | [];
+  setVisibleFeatures: Dispatch<
+    SetStateAction<Feature<MultiPolygon, RampProperties>[] | []>
+  >;
+}) => {
+  if (!visibleFeatures) {
+    visibleFeatures = ramps.features;
+  }
+  const [popupData, setPopupData] = useState<
+    Feature<MultiPolygon, RampProperties> | undefined
+  >(undefined);
+
+  const features: JSX.Element[] = [];
+
+  for (var i = 0; i < ramps.features.length; i++) {
+    var f = ramps.features[i];
+
+    features.push(
+      <Marker
+        key={f.id}
+        latitude={f.geometry.coordinates[0][0][0][1]}
+        longitude={f.geometry.coordinates[0][0][0][0]}
+        onClick={() => {
+          setPopupData(f);
+        }}
+      />
+    );
+  }
```

Comment:
1. There is no need to use the state to render markers. You can directly render markers from props
2. For iterating over a set of data in order to render the UI, prefer the standard of using `.map()`
3. Always encapsulate to its own closure. i.e. move the iteration logic into its own function
4. There is a bug. The popup will not render when you click on the marker. This is due to event bubbling. Best to disable the event from propagating.

```suggestion
   const renderFeatures = () => {
    return (
      <>
        {props.ramps.features.map((feature) => (
          <Marker
            key={feature.id}
            latitude={feature.geometry.coordinates[0][0][0][1]}
            longitude={feature.geometry.coordinates[0][0][0][0]}
            onClick={(e) => {
              e.originalEvent.stopPropagation();
              setPopupData(feature);
            }}
          />
        ))}
      </>
    )
  }
```